diff --git a/compiler/core/Data/DataSystem.v b/compiler/core/Data/DataSystem.v
index c0fc0f4c..af9efa71 100644
--- a/compiler/core/Data/DataSystem.v
+++ b/compiler/core/Data/DataSystem.v
@@ -17,9 +17,9 @@ Require Export DataTypes.
 
 Class basic_model : Type
   := mk_basic_model {
-         basic_model_runtime :> foreign_runtime
-         ; basic_model_foreign_type :> foreign_type
-         ; basic_model_brand_model :> brand_model
-         ; basic_model_foreign_typing :> foreign_typing
+         #[global] basic_model_runtime :: foreign_runtime
+         ; #[global] basic_model_foreign_type :: foreign_type
+         ; #[global] basic_model_brand_model :: brand_model
+         ; #[global] basic_model_foreign_typing :: foreign_typing
        }.
 
diff --git a/compiler/core/Data/ForeignRuntime.v b/compiler/core/Data/ForeignRuntime.v
index 74eba5c9..a990b183 100644
--- a/compiler/core/Data/ForeignRuntime.v
+++ b/compiler/core/Data/ForeignRuntime.v
@@ -17,7 +17,7 @@ Require Export ForeignOperators.
 
 Class foreign_runtime : Type
   := mk_foreign_runtime {
-         foreign_runtime_data :> foreign_data
-         ; foreign_runtime_operators :> foreign_operators
+         #[global] foreign_runtime_data :: foreign_data
+         ; #[global] foreign_runtime_operators :: foreign_operators
        }.
 
diff --git a/compiler/core/Data/ForeignTyping.v b/compiler/core/Data/ForeignTyping.v
index 9b6db32d..5c0a1efb 100644
--- a/compiler/core/Data/ForeignTyping.v
+++ b/compiler/core/Data/ForeignTyping.v
@@ -23,7 +23,7 @@ Class foreign_typing
       {model:brand_model}
   : Type
   := mk_foreign_typing {
-          foreign_typing_data :> foreign_data_typing
-         ; foreign_typing_operators :> foreign_operators_typing
+          #[global] foreign_typing_data :: foreign_data_typing
+         ; #[global] foreign_typing_operators :: foreign_operators_typing
        }.
 
diff --git a/compiler/core/Data/Model/ForeignData.v b/compiler/core/Data/Model/ForeignData.v
index 47f21a7f..8f719f4c 100644
--- a/compiler/core/Data/Model/ForeignData.v
+++ b/compiler/core/Data/Model/ForeignData.v
@@ -22,14 +22,14 @@ Section ForeignData.
            foreign_data_model : Set
 (* TODO: #35 is needed before we can generalize this
   ; foreign_data_equiv (a b : foreign_data_model) : Prop
-  ; foreign_data_equiv_equiv :> Equivalence foreign_data_equiv
+  ; #[global] foreign_data_equiv_equiv :: Equivalence foreign_data_equiv
  *) 
-           ; foreign_data_dec :> EqDec foreign_data_model eq
+           ; #[global] foreign_data_dec :: EqDec foreign_data_model eq
            ; foreign_data_normalized (a : foreign_data_model) : Prop
            ; foreign_data_normalize (a : foreign_data_model) : foreign_data_model
            ; foreign_data_normalize_normalizes (a : foreign_data_model) : foreign_data_normalized (foreign_data_normalize a)
            ; foreign_data_normalize_idempotent (a : foreign_data_model) : foreign_data_normalized a -> foreign_data_normalize a = a
-           ; foreign_data_tostring :> ToString foreign_data_model
+           ; #[global] foreign_data_tostring :: ToString foreign_data_model
       }.
 
 End ForeignData.
diff --git a/compiler/core/Data/Operators/ForeignOperators.v b/compiler/core/Data/Operators/ForeignOperators.v
index 64c5818d..f87a6252 100644
--- a/compiler/core/Data/Operators/ForeignOperators.v
+++ b/compiler/core/Data/Operators/ForeignOperators.v
@@ -26,8 +26,8 @@ Section ForeignOperators.
     := mk_foreign_operators {
            (* Unary operators *)
            foreign_operators_unary : Set
-           ; foreign_operators_unary_dec :> EqDec foreign_operators_unary eq
-           ; foreign_operators_unary_tostring :> ToString foreign_operators_unary
+           ; #[global] foreign_operators_unary_dec :: EqDec foreign_operators_unary eq
+           ; #[global] foreign_operators_unary_tostring :: ToString foreign_operators_unary
            ; foreign_operators_unary_interp
                (br:brand_relation_t)
                (op:foreign_operators_unary)
@@ -40,8 +40,8 @@ Section ForeignOperators.
                data_normalized br o
            (* Binary operators *)
            ; foreign_operators_binary : Set
-           ; foreign_operators_binary_dec :> EqDec foreign_operators_binary eq
-           ; foreign_operators_binary_tostring :> ToString foreign_operators_binary
+           ; #[global] foreign_operators_binary_dec :: EqDec foreign_operators_binary eq
+           ; #[global] foreign_operators_binary_tostring :: ToString foreign_operators_binary
            ; foreign_operators_binary_interp
                (br:brand_relation_t)
                (op:foreign_operators_binary)
diff --git a/compiler/core/EJson/Model/ForeignEJson.v b/compiler/core/EJson/Model/ForeignEJson.v
index 2dfd7a7a..8e56947b 100644
--- a/compiler/core/EJson/Model/ForeignEJson.v
+++ b/compiler/core/EJson/Model/ForeignEJson.v
@@ -20,12 +20,12 @@ Section ForeignEJson.
   
   Class foreign_ejson (foreign_ejson_model : Set) : Type
     := mk_foreign_ejson {
-           foreign_ejson_dec :> EqDec foreign_ejson_model eq
+           #[global] foreign_ejson_dec :: EqDec foreign_ejson_model eq
            ; foreign_ejson_normalized (a : foreign_ejson_model) : Prop
            ; foreign_ejson_normalize (a : foreign_ejson_model) : foreign_ejson_model
            ; foreign_ejson_normalize_normalizes (a : foreign_ejson_model) : foreign_ejson_normalized (foreign_ejson_normalize a)
            ; foreign_ejson_normalize_idempotent (a : foreign_ejson_model) : foreign_ejson_normalized a -> foreign_ejson_normalize a = a
-           ; foreign_ejson_tostring :> ToString foreign_ejson_model
+           ; #[global] foreign_ejson_tostring :: ToString foreign_ejson_model
       }.
 
 End ForeignEJson.
diff --git a/compiler/core/EJson/Operators/ForeignEJsonRuntime.v b/compiler/core/EJson/Operators/ForeignEJsonRuntime.v
index 0210bf24..f2bfffed 100644
--- a/compiler/core/EJson/Operators/ForeignEJsonRuntime.v
+++ b/compiler/core/EJson/Operators/ForeignEJsonRuntime.v
@@ -26,8 +26,8 @@ Section ForeignEJsonRuntime.
         {fejson:foreign_ejson foreign_ejson_model}
   : Type
     := mk_foreign_ejson_runtime {
-           foreign_ejson_runtime_op_dec :> EqDec foreign_ejson_runtime_op eq
-           ; foreign_ejson_runtime_op_tostring :> ToString foreign_ejson_runtime_op
+           #[global] foreign_ejson_runtime_op_dec :: EqDec foreign_ejson_runtime_op eq
+           ; #[global] foreign_ejson_runtime_op_tostring :: ToString foreign_ejson_runtime_op
            ; foreign_ejson_runtime_op_interp 
                (f:foreign_ejson_runtime_op)
                (dl:list (@ejson foreign_ejson_model)) : option (@ejson foreign_ejson_model)
diff --git a/compiler/core/Imp/Lang/Imp.v b/compiler/core/Imp/Lang/Imp.v
index fd120929..91fec477 100644
--- a/compiler/core/Imp/Lang/Imp.v
+++ b/compiler/core/Imp/Lang/Imp.v
@@ -21,7 +21,6 @@ Require Import ZArith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/Imp/Lang/ImpDataEval.v b/compiler/core/Imp/Lang/ImpDataEval.v
index 0c157d8a..91434bc6 100644
--- a/compiler/core/Imp/Lang/ImpDataEval.v
+++ b/compiler/core/Imp/Lang/ImpDataEval.v
@@ -19,7 +19,6 @@ Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
 Require Import ZArith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/Imp/Lang/ImpEJsonEval.v b/compiler/core/Imp/Lang/ImpEJsonEval.v
index 29cf6de5..33a20ff1 100644
--- a/compiler/core/Imp/Lang/ImpEJsonEval.v
+++ b/compiler/core/Imp/Lang/ImpEJsonEval.v
@@ -19,7 +19,6 @@ Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
 Require Import ZArith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/Imp/Lang/ImpVars.v b/compiler/core/Imp/Lang/ImpVars.v
index 165cae1b..fe685b35 100644
--- a/compiler/core/Imp/Lang/ImpVars.v
+++ b/compiler/core/Imp/Lang/ImpVars.v
@@ -19,7 +19,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/LambdaNRA/Typing/TLambdaNRA.v b/compiler/core/LambdaNRA/Typing/TLambdaNRA.v
index 70b5a170..cf745d12 100644
--- a/compiler/core/LambdaNRA/Typing/TLambdaNRA.v
+++ b/compiler/core/LambdaNRA/Typing/TLambdaNRA.v
@@ -173,7 +173,6 @@ Section TLambdaNRA.
         apply some_lift in edd1.
         destruct edd1 as [dd11 ? tdd11]; subst.
         rewrite e.
-        rewrite lnra_lambda_eval_lambda_eq.
         assert (bt1:bindings_type (rec_sort (env ++ [(s, a)])) (rec_sort (Γ ++ [(s, τ₀)])) ).
         {
           apply bindings_type_sort.
@@ -207,8 +206,6 @@ Section TLambdaNRA.
           constructor; simpl; intuition. 
         }
         destruct (IHe1 _ _ _ bt1 H1) as [dd3 [edd3 tdd3]].
-        rewrite (omap_product_ext _  (fun d => lambda_nra_eval brand_relation_brands cenv (rec_sort (env ++ [(s, d)])) e1));
-          [ | intros; apply lnra_lambda_eval_lambda_eq ].
         dtype_inverter.
         apply Col_inv in tdd1.
         apply Col_inv in tdd3.
@@ -297,7 +294,6 @@ Section TLambdaNRA.
         apply some_lift in edd1.
         destruct edd1 as [dd11 dd11e dd11t]; subst.
         rewrite dd11e.
-        rewrite lnra_lambda_eval_lambda_eq.
         assert (bt1:bindings_type (rec_sort (env ++ [(s, a)])) (rec_sort (Γ ++ [(s, τ0)])) ).
         {
           apply bindings_type_sort.
diff --git a/compiler/core/NNRC/Lang/NNRC.v b/compiler/core/NNRC/Lang/NNRC.v
index 3b8520f7..78db3327 100644
--- a/compiler/core/NNRC/Lang/NNRC.v
+++ b/compiler/core/NNRC/Lang/NNRC.v
@@ -38,7 +38,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRC/Lang/NNRCStratify.v b/compiler/core/NNRC/Lang/NNRCStratify.v
index 3de34849..b10dd105 100644
--- a/compiler/core/NNRC/Lang/NNRCStratify.v
+++ b/compiler/core/NNRC/Lang/NNRCStratify.v
@@ -1283,7 +1283,6 @@ Section Stratify.
       unfold nnrc_eval in *.
       destruct a; simpl in *.
       match_case; intros ? eqq; simpl.
-      rewrite IHsdefs; trivial.
     Qed.
 
     Lemma eval_substs_app h cenv (sdefs1 sdefs2:list (var*nnrc)) (env:bindings) :
diff --git a/compiler/core/NNRC/Lang/NNRCSugar.v b/compiler/core/NNRC/Lang/NNRCSugar.v
index a69e5742..3f3b8e97 100644
--- a/compiler/core/NNRC/Lang/NNRCSugar.v
+++ b/compiler/core/NNRC/Lang/NNRCSugar.v
@@ -20,7 +20,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRC/Optim/TNNRCRewrite.v b/compiler/core/NNRC/Optim/TNNRCRewrite.v
index 472b042a..07d3aa7a 100644
--- a/compiler/core/NNRC/Optim/TNNRCRewrite.v
+++ b/compiler/core/NNRC/Optim/TNNRCRewrite.v
@@ -1218,8 +1218,7 @@ Section TNNRCRewrite.
     - econstructor; eauto 2.
       econstructor; eauto 2.
       + reflexivity.
-      + destruct (@in_dec string string_dec
-                        s sl); [| intuition]; [intuition | ].
+      + destruct (@in_dec string string_dec s sl); [intuition | ].
         simpl.
         econstructor; eauto.
     - intros.
diff --git a/compiler/core/NNRCMR/Lang/ForeignReduceOps.v b/compiler/core/NNRCMR/Lang/ForeignReduceOps.v
index 5a3ba607..4bc0d0c4 100644
--- a/compiler/core/NNRCMR/Lang/ForeignReduceOps.v
+++ b/compiler/core/NNRCMR/Lang/ForeignReduceOps.v
@@ -22,8 +22,8 @@ Section ForeignReduceOps.
   : Type
     := mk_foreign_reduce_op {
            foreign_reduce_op_type : Set
-           ; foreign_reduce_op_dec :> EqDec foreign_reduce_op_type eq
-           ; foreign_reduce_op_tostring :> ToString foreign_reduce_op_type
+           ; #[global] foreign_reduce_op_dec :: EqDec foreign_reduce_op_type eq
+           ; #[global] foreign_reduce_op_tostring :: ToString foreign_reduce_op_type
            ; foreign_reduce_op_interp 
                (br:brand_relation_t)
                (op:foreign_reduce_op_type)
diff --git a/compiler/core/NNRS/Lang/NNRS.v b/compiler/core/NNRS/Lang/NNRS.v
index 33cfdc91..275fa7ec 100644
--- a/compiler/core/NNRS/Lang/NNRS.v
+++ b/compiler/core/NNRS/Lang/NNRS.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Lang/NNRSCrossShadow.v b/compiler/core/NNRS/Lang/NNRSCrossShadow.v
index d7aa8110..98091162 100644
--- a/compiler/core/NNRS/Lang/NNRSCrossShadow.v
+++ b/compiler/core/NNRS/Lang/NNRSCrossShadow.v
@@ -27,7 +27,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Lang/NNRSEval.v b/compiler/core/NNRS/Lang/NNRSEval.v
index 22426879..c3abfd88 100644
--- a/compiler/core/NNRS/Lang/NNRSEval.v
+++ b/compiler/core/NNRS/Lang/NNRSEval.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Lang/NNRSRename.v b/compiler/core/NNRS/Lang/NNRSRename.v
index 8f500788..3d6ab02f 100644
--- a/compiler/core/NNRS/Lang/NNRSRename.v
+++ b/compiler/core/NNRS/Lang/NNRSRename.v
@@ -18,7 +18,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Lang/NNRSSem.v b/compiler/core/NNRS/Lang/NNRSSem.v
index 72a2d007..9892ac1c 100644
--- a/compiler/core/NNRS/Lang/NNRSSem.v
+++ b/compiler/core/NNRS/Lang/NNRSSem.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Lang/NNRSSemEval.v b/compiler/core/NNRS/Lang/NNRSSemEval.v
index d9634343..4dcb5b07 100644
--- a/compiler/core/NNRS/Lang/NNRSSemEval.v
+++ b/compiler/core/NNRS/Lang/NNRSSemEval.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Lang/NNRSVars.v b/compiler/core/NNRS/Lang/NNRSVars.v
index da582b92..910c95bb 100644
--- a/compiler/core/NNRS/Lang/NNRSVars.v
+++ b/compiler/core/NNRS/Lang/NNRSVars.v
@@ -18,7 +18,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Typing/TNNRSCrossShadow.v b/compiler/core/NNRS/Typing/TNNRSCrossShadow.v
index 1a7c82d9..9113b50c 100644
--- a/compiler/core/NNRS/Typing/TNNRSCrossShadow.v
+++ b/compiler/core/NNRS/Typing/TNNRSCrossShadow.v
@@ -21,7 +21,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRS/Typing/TNNRSRename.v b/compiler/core/NNRS/Typing/TNNRSRename.v
index 3ba636a1..a3363bef 100644
--- a/compiler/core/NNRS/Typing/TNNRSRename.v
+++ b/compiler/core/NNRS/Typing/TNNRSRename.v
@@ -17,7 +17,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Lang/NNRSimp.v b/compiler/core/NNRSimp/Lang/NNRSimp.v
index f1e0af56..67f804e4 100644
--- a/compiler/core/NNRSimp/Lang/NNRSimp.v
+++ b/compiler/core/NNRSimp/Lang/NNRSimp.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Lang/NNRSimpEval.v b/compiler/core/NNRSimp/Lang/NNRSimpEval.v
index c85a8853..456e27f1 100644
--- a/compiler/core/NNRSimp/Lang/NNRSimpEval.v
+++ b/compiler/core/NNRSimp/Lang/NNRSimpEval.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Lang/NNRSimpRename.v b/compiler/core/NNRSimp/Lang/NNRSimpRename.v
index 94ae8285..0355fa4c 100644
--- a/compiler/core/NNRSimp/Lang/NNRSimpRename.v
+++ b/compiler/core/NNRSimp/Lang/NNRSimpRename.v
@@ -18,7 +18,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Lang/NNRSimpSem.v b/compiler/core/NNRSimp/Lang/NNRSimpSem.v
index 875ffb2b..3b3bb104 100644
--- a/compiler/core/NNRSimp/Lang/NNRSimpSem.v
+++ b/compiler/core/NNRSimp/Lang/NNRSimpSem.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Lang/NNRSimpSemEval.v b/compiler/core/NNRSimp/Lang/NNRSimpSemEval.v
index e867b6af..07f37e14 100644
--- a/compiler/core/NNRSimp/Lang/NNRSimpSemEval.v
+++ b/compiler/core/NNRSimp/Lang/NNRSimpSemEval.v
@@ -23,7 +23,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Lang/NNRSimpVars.v b/compiler/core/NNRSimp/Lang/NNRSimpVars.v
index cd831368..3efbb9c1 100644
--- a/compiler/core/NNRSimp/Lang/NNRSimpVars.v
+++ b/compiler/core/NNRSimp/Lang/NNRSimpVars.v
@@ -19,7 +19,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NNRSimp/Typing/TNNRSimpRename.v b/compiler/core/NNRSimp/Typing/TNNRSimpRename.v
index 340111b5..a78d8889 100644
--- a/compiler/core/NNRSimp/Typing/TNNRSimpRename.v
+++ b/compiler/core/NNRSimp/Typing/TNNRSimpRename.v
@@ -18,7 +18,6 @@ Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Arith.
-Require Import Max.
 Require Import Bool.
 Require Import Peano_dec.
 Require Import EquivDec.
diff --git a/compiler/core/NRA/Context/NRAContext.v b/compiler/core/NRA/Context/NRAContext.v
index fd5b15a5..44c4cbea 100644
--- a/compiler/core/NRA/Context/NRAContext.v
+++ b/compiler/core/NRA/Context/NRAContext.v
@@ -24,7 +24,6 @@ Require Import Program.
 Require Import List.
 Require Import Permutation.
 Require Import String.
-Require Import NPeano.
 Require Import Arith.
 Require Import Bool.
 Require Import Utils.
diff --git a/compiler/core/NRA/Optim/NRARewriteContext.v b/compiler/core/NRA/Optim/NRARewriteContext.v
index 2802be2b..dfb5ab40 100644
--- a/compiler/core/NRA/Optim/NRARewriteContext.v
+++ b/compiler/core/NRA/Optim/NRARewriteContext.v
@@ -22,7 +22,6 @@ Require Import Setoid.
 Require Import EquivDec.
 Require Import Program.
 Require Import Arith.
-Require Import NPeano.
 Require Import Lia.
 Require Import List.
 Require Import Utils.
diff --git a/compiler/core/NRA/Optim/TNRARewrite.v b/compiler/core/NRA/Optim/TNRARewrite.v
index 1d460e90..26da0ee0 100644
--- a/compiler/core/NRA/Optim/TNRARewrite.v
+++ b/compiler/core/NRA/Optim/TNRARewrite.v
@@ -74,7 +74,7 @@ Section TNRARewrite.
     brand_relation_brands ⊢ (σ⟨ op2 ∧ op1 ⟩(op)) @ₐ x ⊣ c.
   Proof.
     intros ? ? ? Hcenv; intros; simpl.
-    assert (exists d, (brand_relation_brands ⊢ op@ₐx ⊣ c = Some d /\ (d ▹ (Coll τ))))
+    assert (exists d, (brand_relation_brands ⊢ op @ₐ x ⊣ c = Some d /\ (d ▹ (Coll τ))))
       by (apply (@typed_nra_yields_typed_data m τc τin); assumption).
     elim H3; clear H3; intros.
     elim H3; clear H3; intros.
@@ -94,7 +94,7 @@ Section TNRARewrite.
     rewrite Forall_forall in H6; simpl in H6.
     assert (data_type a τ)
       by intuition.
-    assert (exists d, (brand_relation_brands ⊢ op2@ₐa ⊣ c = Some d /\ (d ▹ Bool)))
+    assert (exists d, (brand_relation_brands ⊢ op2 @ₐ a ⊣ c = Some d /\ (d ▹ Bool)))
       by (apply (@typed_nra_yields_typed_data m τc τ); assumption).
     destruct H4 as [? [eqq dt]].
     rewrite eqq; clear eqq.
@@ -108,13 +108,13 @@ Section TNRARewrite.
           end) dl).
     - destruct x1.
       + simpl.
-        assert (exists d, (brand_relation_brands ⊢ op1@ₐa ⊣ c = Some d /\ (d ▹ Bool)))
+        assert (exists d, (brand_relation_brands ⊢ op1 @ₐ a ⊣ c = Some d /\ (d ▹ Bool)))
           by (apply (@typed_nra_yields_typed_data m τc τ); assumption).
         destruct H4 as [? [eqq dt]].
         rewrite eqq; clear eqq.
         dtype_inverter.
         reflexivity.
-      + assert (exists d, (brand_relation_brands ⊢ op1@ₐa ⊣ c = Some d /\ (d ▹ Bool)))
+      + assert (exists d, (brand_relation_brands ⊢ op1 @ₐ a ⊣ c = Some d /\ (d ▹ Bool)))
           by (apply (@typed_nra_yields_typed_data m τc τ); assumption).
         destruct H4 as [? [eqq dt]].
         rewrite eqq; clear eqq.
@@ -125,7 +125,7 @@ Section TNRARewrite.
       | Some (dbool b0) => Some b0
       | _ => None
       end) l); reflexivity.
-    - destruct (brand_relation_brands ⊢ op1@ₐa ⊣ c); try reflexivity.
+    - destruct (brand_relation_brands ⊢ op1 @ₐ a ⊣ c); try reflexivity.
       destruct d; reflexivity.
   Qed.      
 
@@ -234,7 +234,7 @@ Section TNRARewrite.
   Defined.
 
   Lemma typed_nra_total_bool_consistent {τc} {τ} c (op:τ ⇝ Bool ⊣ τc) (d: {x:data|(x ▹ τ)}) (Hcenv:bindings_type c τc) :
-    match (brand_relation_brands ⊢ `op@ₐ`d ⊣ c) with
+    match (brand_relation_brands ⊢ `op @ₐ `d ⊣ c) with
       | Some (dbool b) => Some b
       | _ => None
     end = Some (typed_nra_total_bool c op Hcenv d).
@@ -252,7 +252,7 @@ Section TNRARewrite.
   Qed.
 
   Lemma typed_nra_total_bool_consistent2 {τc} {τ} c (op:τ ⇝ Bool ⊣ τc) (x:data) (pf:(x ▹ τ)) (Hcenv:bindings_type c τc) :
-    match (brand_relation_brands ⊢ `op@ₐx ⊣ c) with
+    match (brand_relation_brands ⊢ `op @ₐ x ⊣ c) with
       | Some (dbool b) => Some b
       | _ => None
     end = Some (typed_nra_total_bool c op Hcenv (exist _ x pf)).
@@ -265,7 +265,7 @@ Section TNRARewrite.
     data_type d τ ->
     exists b' : bool,
       (fun x' : data =>
-         match brand_relation_brands ⊢ (`op)@ₐ x' ⊣ c with
+         match brand_relation_brands ⊢ (`op) @ₐ x' ⊣ c with
            | Some (dbool b) => Some b
            | _ => None
          end) d = Some b'.
@@ -389,9 +389,9 @@ Section TNRARewrite.
     unfold tnra_eq.
     intros ? ? x c ? Hcenv.
     rewrite H; rewrite H0; clear H H0 opl opr; simpl.
-    assert (exists d1, (brand_relation_brands ⊢ `op1@ₐx ⊣ c = Some d1 /\ (d1 ▹ (Coll τ))))
+    assert (exists d1, (brand_relation_brands ⊢ `op1 @ₐ x ⊣ c = Some d1 /\ (d1 ▹ (Coll τ))))
       by (apply (@typed_nra_yields_typed_data m τc τin); [assumption|assumption|apply (proj2_sig op1)]).
-    assert (exists d2, (brand_relation_brands ⊢ `op2@ₐx ⊣ c = Some d2 /\ (d2 ▹ (Coll τ))))
+    assert (exists d2, (brand_relation_brands ⊢ `op2 @ₐ x ⊣ c = Some d2 /\ (d2 ▹ (Coll τ))))
       by (apply (@typed_nra_yields_typed_data m τc τin); [assumption|assumption|apply (proj2_sig op2)]).
     elim H; elim H0; clear H H0; intros.
     elim H; elim H0; clear H H0; intros.
diff --git a/compiler/core/NRA/Typing/TNRA.v b/compiler/core/NRA/Typing/TNRA.v
index 9778d5c2..52d996d0 100644
--- a/compiler/core/NRA/Typing/TNRA.v
+++ b/compiler/core/NRA/Typing/TNRA.v
@@ -319,7 +319,7 @@ Section TNRA.
                 data_type d (Rec Closed τ₁ pf1) ->
                 exists x : data,
                    brand_relation_brands ⊢ op1 @ₐ y  ⊣ c = Some x /\ data_type x (Coll (Rec Closed τ₂ pf2))) ->
-    exists x : list data, (omap_product (fun z =>  brand_relation_brands ⊢ op1@ₐ y ⊣ c) dl = Some x) /\ data_type (dcoll x) (Coll (Rec Closed τ₃ pf3)).
+    exists x : list data, (omap_product (fun z =>  brand_relation_brands ⊢ op1 @ₐ y ⊣ c) dl = Some x) /\ data_type (dcoll x) (Coll (Rec Closed τ₃ pf3)).
   Proof.
     intros Hcenv.
     intros; rewrite Forall_forall in *.
@@ -521,7 +521,7 @@ Section TNRA.
     intros HdT.
     generalize (typed_nra_yields_typed_data c d op dt_c HdT HOpT).
     intros.
-    destruct (brand_relation_brands ⊢ op@ₐd ⊣ c).
+    destruct (brand_relation_brands ⊢ op @ₐ d ⊣ c).
     assert (data_type d0 τout).
     - inversion H. inversion H0. inversion H1. trivial.
     - exists d0. trivial.
@@ -532,7 +532,7 @@ Section TNRA.
   Lemma typed_nra_total_eq_matches_eval {τc} {τin τout} (op:nra) (HOpT: op ▷ τin >=> τout ⊣ τc) c (d:data)
     (dt_c: bindings_type c τc)
         (pf: data_type d τin) :
-    (brand_relation_brands ⊢ op @ₐd ⊣ c) = Some (proj1_sig (typed_nra_total op HOpT c d dt_c pf)).
+    (brand_relation_brands ⊢ op @ₐ d ⊣ c) = Some (proj1_sig (typed_nra_total op HOpT c d dt_c pf)).
   Proof.
     unfold typed_nra_total.
     generalize (typed_nra_yields_typed_data c d op dt_c pf HOpT); intros.
diff --git a/compiler/core/Tests/LambdaNRATest.v b/compiler/core/Tests/LambdaNRATest.v
index 2497a979..b9704905 100644
--- a/compiler/core/Tests/LambdaNRATest.v
+++ b/compiler/core/Tests/LambdaNRATest.v
@@ -109,8 +109,7 @@ Section LambdaNRATest.
                   (String (Ascii.Ascii false false false false true true true false)
                      EmptyString)))); simpl; intros.
     - unfold lambda_nra_eval in H |- *.
-      rewrite H; clear H.
-      rewrite olift_some.
+      unfold olift in H; rewrite H; clear H.
       match_case_in IHl; intros; rewrite H in IHl.
       + apply some_lift in H.
         destruct H as [? ? ?]; subst.
@@ -150,7 +149,7 @@ Section LambdaNRATest.
           rewrite @assoc_lookupr_app.
           simpl.
           trivial.
-        * repeat match_destr.
+        * unfold edot; simpl; rewrite (@assoc_lookupr_drec_sort string ODT_string); rewrite @assoc_lookupr_app; simpl; trivial.
       + symmetry in IHl.
         apply none_lift in IHl.
         apply none_lift in H.
@@ -158,11 +157,7 @@ Section LambdaNRATest.
         simpl.
         destruct d; simpl; trivial.
         match_destr.
-    - match_case; intros.
-      unfold lambda_nra_eval in H, H0.
-      rewrite H in H0.
-      simpl in H0.
-      discriminate.
+    - match_case; intros; try unfold lambda_nra_eval in H, H0; unfold olift in H, H0; rewrite H in H0; simpl in H0; try discriminate; rewrite H; simpl; trivial.
   Qed.
 
   
diff --git a/compiler/core/Translation/Lang/ImpDatatoImpEJson.v b/compiler/core/Translation/Lang/ImpDatatoImpEJson.v
index 858f85ee..5b818d4e 100644
--- a/compiler/core/Translation/Lang/ImpDatatoImpEJson.v
+++ b/compiler/core/Translation/Lang/ImpDatatoImpEJson.v
@@ -496,7 +496,7 @@ Section ImpDatatoImpEJson.
           unfold lift, olift;
           case_eq
             (lift_map (fun x : option imp_data_model => x)
-                      (map (fun x : imp_data_expr => imp_data_expr_eval h σ x) el)); intros;
+                      (map (imp_data_expr_eval h σ) el)); intros;
           unfold imp_data_model, ImpEval.imp_expr, imp_data_expr, imp_data_model, foreign_runtime_data in *;
           rewrite H0; reflexivity].
       (* just one param *)
@@ -663,7 +663,7 @@ Section ImpDatatoImpEJson.
           unfold lift, olift;
           case_eq
             (lift_map (fun x : option imp_data_model => x)
-                      (map (fun x : imp_data_expr => imp_data_expr_eval h σ x) el)); intros;
+                      (map (imp_data_expr_eval h σ) el)); intros;
           unfold imp_data_model, ImpEval.imp_expr, imp_data_expr, imp_data_model, foreign_runtime_data in *;
           rewrite H0; reflexivity].
       (* just two param *)
diff --git a/compiler/core/Translation/Lang/NNRCtoNNRCMR.v b/compiler/core/Translation/Lang/NNRCtoNNRCMR.v
index 69bad4c7..f5b21b37 100644
--- a/compiler/core/Translation/Lang/NNRCtoNNRCMR.v
+++ b/compiler/core/Translation/Lang/NNRCtoNNRCMR.v
@@ -360,6 +360,9 @@ Section NNRCtoNNRCMR.
       simpl.
       case (equiv_dec output output);
         try congruence; intros.
+      destruct (@equiv_dec var (@eq var) (@eq_equivalence var) string_eqdec output output); try congruence.
+      simpl.
+      destruct (equiv_dec output output); [reflexivity | congruence].
     - specialize (H1 v); clear H0.
       simpl in *.
       dest_eqdec; try congruence.
@@ -410,8 +413,9 @@ Section NNRCtoNNRCMR.
       case (equiv_dec output output);
         try congruence; intros.
       simpl.
-      case (equiv_dec output output);
-        congruence.
+      destruct (@equiv_dec var (@eq var) (@eq_equivalence var) string_eqdec output output); try congruence.
+      simpl.
+      destruct (equiv_dec output output); [reflexivity | congruence].
   Qed.
 
   (* --------------------------------- *
diff --git a/compiler/core/Translation/Model/ForeignDataToEJson.v b/compiler/core/Translation/Model/ForeignDataToEJson.v
index 856190f7..d0c0ebaf 100644
--- a/compiler/core/Translation/Model/ForeignDataToEJson.v
+++ b/compiler/core/Translation/Model/ForeignDataToEJson.v
@@ -30,7 +30,7 @@ Class foreign_to_ejson
       {fruntime:foreign_runtime}
   : Type
   := mk_foreign_to_ejson {
-         foreign_to_ejson_runtime :> foreign_ejson_runtime foreign_ejson_runtime_op
+         #[global] foreign_to_ejson_runtime :: foreign_ejson_runtime foreign_ejson_runtime_op
          ; foreign_to_ejson_to_data
              (j:foreign_ejson_model) : foreign_data_model
          ; foreign_to_ejson_from_data
diff --git a/compiler/core/Translation/Typing/TNRAtocNNRC.v b/compiler/core/Translation/Typing/TNRAtocNNRC.v
index ef5e2873..c40444ce 100644
--- a/compiler/core/Translation/Typing/TNRAtocNNRC.v
+++ b/compiler/core/Translation/Typing/TNRAtocNNRC.v
@@ -110,7 +110,6 @@ Section TNRAtocNNRC.
       + econstructor. econstructor; eauto.
         econstructor; eauto.
         econstructor; eauto.
-        eapply Forall_nil.
       + apply IHnra_type2; simpl; trivial; match_destr; elim_fresh e.
       + econstructor; eauto.
         simpl; match_destr; elim_fresh e.
diff --git a/compiler/core/Translation/Typing/TcNRAEnvtocNNRC.v b/compiler/core/Translation/Typing/TcNRAEnvtocNNRC.v
index 2b95aa04..ce4b841e 100644
--- a/compiler/core/Translation/Typing/TcNRAEnvtocNNRC.v
+++ b/compiler/core/Translation/Typing/TcNRAEnvtocNNRC.v
@@ -109,7 +109,6 @@ Section TcNRAEnvtocNNRC.
       + econstructor. econstructor; qeauto.
         econstructor; qeauto.
         econstructor; qeauto.
-        eapply Forall_nil.
       + apply IHnraenv_core_type2; simpl; trivial; match_destr; elim_fresh e.
       + econstructor; qeauto.
         simpl; match_destr; elim_fresh e.
diff --git a/compiler/core/TypeSystem/ForeignType.v b/compiler/core/TypeSystem/ForeignType.v
index 42647bdc..4e2300da 100644
--- a/compiler/core/TypeSystem/ForeignType.v
+++ b/compiler/core/TypeSystem/ForeignType.v
@@ -22,16 +22,16 @@ Section ForeignType.
            foreign_type_type : Set
 (* TODO: #35 is needed before we can generalize this
   ; foreign_type_equiv (a b : foreign_type_type) : Prop
-  ; foreign_type_equiv_equiv :> Equivalence foreign_type_equiv
+  ; #[global] foreign_type_equiv_equiv :: Equivalence foreign_type_equiv
  *) 
-           ; foreign_type_dec :> EqDec foreign_type_type eq
-           ; foreign_type_lattice :> Lattice foreign_type_type eq
+           ; #[global] foreign_type_dec :: EqDec foreign_type_type eq
+           ; #[global] foreign_type_lattice :: Lattice foreign_type_type eq
            ; foreign_type_sub (a b : foreign_type_type) : Prop
            ; foreign_type_sub_dec (a b : foreign_type_type) :
                { foreign_type_sub a b} + {~ foreign_type_sub a b}
-           ; foreign_type_sub_pre :> PreOrder foreign_type_sub
-           ; foreign_type_sub_part :> PartialOrder eq foreign_type_sub
-           ; foreign_type_olattice :> OLattice eq foreign_type_sub
+           ; #[global] foreign_type_sub_pre :: PreOrder foreign_type_sub
+           ; #[global] foreign_type_sub_part :: PartialOrder eq foreign_type_sub
+           ; #[global] foreign_type_olattice :: OLattice eq foreign_type_sub
       }.
   
 End ForeignType.
diff --git a/compiler/core/TypeSystem/TBrandModel.v b/compiler/core/TypeSystem/TBrandModel.v
index def4ec1c..116fd2b3 100644
--- a/compiler/core/TypeSystem/TBrandModel.v
+++ b/compiler/core/TypeSystem/TBrandModel.v
@@ -124,8 +124,8 @@ Section TBrandModel.
   Class brand_model :=
     mkBrand_model {
         (* note that brand_context_brand_model can have duplicates in the domain *)
-        brand_model_relation :> brand_relation;
-        brand_model_context :> brand_context;
+        #[global] brand_model_relation :: brand_relation;
+        #[global] brand_model_context :: brand_context;
         brand_model_domain_b :
          holds (brand_model_domain_dec);
             
diff --git a/compiler/core/Updates/UPropag.v b/compiler/core/Updates/UPropag.v
index 6af99c87..f63c5d1b 100644
--- a/compiler/core/Updates/UPropag.v
+++ b/compiler/core/Updates/UPropag.v
@@ -397,7 +397,7 @@ Section UPropag.
       revert H.
       generalize (mult s a) as ns.
       intros.
-      rewrite Max.max_idempotent.
+      rewrite Nat.max_idempotent.
       lia.
   Qed.
 
diff --git a/compiler/core/Utils/Bag.v b/compiler/core/Utils/Bag.v
index 65476395..122478f6 100644
--- a/compiler/core/Utils/Bag.v
+++ b/compiler/core/Utils/Bag.v
@@ -15,8 +15,6 @@
 (** This module provides support for bags (or multisets). *)
 
 Require Import Arith.
-Require Import Min.
-Require Import Max.
 Require Import ZArith.
 Require Import Lia.
 Require Import Permutation.
diff --git a/compiler/core/Utils/Bindings.v b/compiler/core/Utils/Bindings.v
index 0ad2b75a..a2118b1c 100644
--- a/compiler/core/Utils/Bindings.v
+++ b/compiler/core/Utils/Bindings.v
@@ -40,9 +40,9 @@ Require Import StringAdd.
 Section Bindings.
 
   Class ODT {K:Type}
-    := mkODT { ODT_eqdec:>EqDec K eq;
+    := mkODT { #[global] ODT_eqdec ::EqDec K eq;
                ODT_lt:K -> K -> Prop;
-               ODT_lt_strorder:>StrictOrder ODT_lt;
+               #[global] ODT_lt_strorder ::StrictOrder ODT_lt;
                ODT_lt_dec: forall (a b:K), {ODT_lt a b} + {~ODT_lt a b};
                ODT_compare:K -> K -> comparison;
                ODT_compare_spec: forall x y : K,
diff --git a/compiler/core/Utils/BindingsNat.v b/compiler/core/Utils/BindingsNat.v
index aeb07a28..333f5737 100644
--- a/compiler/core/Utils/BindingsNat.v
+++ b/compiler/core/Utils/BindingsNat.v
@@ -16,7 +16,6 @@
 natural number. *)
 
 Require Import Arith.
-Require Import NPeano.
 Require Import Bindings.
 
 Section BindingsNat.
diff --git a/compiler/core/Utils/CoqLibAdd.v b/compiler/core/Utils/CoqLibAdd.v
index d7547576..d6e7764d 100644
--- a/compiler/core/Utils/CoqLibAdd.v
+++ b/compiler/core/Utils/CoqLibAdd.v
@@ -28,7 +28,6 @@ Require Import EquivDec.
 Require Import Equivalence.
 Require Import Peano_dec.
 Require Import ZArith.
-Require Import Zdigits.
 Require Import Znat.
 Require Import Recdef.
 Require Import Compare_dec.
diff --git a/compiler/core/Utils/Fresh.v b/compiler/core/Utils/Fresh.v
index 59c1ae7d..76895742 100644
--- a/compiler/core/Utils/Fresh.v
+++ b/compiler/core/Utils/Fresh.v
@@ -18,7 +18,7 @@ as strings). *)
 Require Import String.
 Require Import List.
 Require Import Permutation.
-Require Import Arith Min.
+Require Import Arith.
 Require Import EquivDec.
 Require Import Morphisms.
 Require Import Lia.
diff --git a/compiler/core/Utils/Lattice.v b/compiler/core/Utils/Lattice.v
index d6f1e5cd..41b7a8f6 100644
--- a/compiler/core/Utils/Lattice.v
+++ b/compiler/core/Utils/Lattice.v
@@ -52,8 +52,8 @@ Section Lattice.
         meet : A -> A -> A;
         join : A -> A -> A;
 
-        meet_morphism :> Proper (eqA ==> eqA ==> eqA) meet ;
-        join_morphism :> Proper (eqA ==> eqA ==> eqA) join ;
+        #[global] meet_morphism :: Proper (eqA ==> eqA ==> eqA) meet ;
+        #[global] join_morphism :: Proper (eqA ==> eqA ==> eqA) join ;
         
         meet_commutative : commutative eqA meet;
         meet_associative    : associative eqA meet;
diff --git a/compiler/core/Utils/LiftIterators.v b/compiler/core/Utils/LiftIterators.v
index c0880c52..b453eba7 100644
--- a/compiler/core/Utils/LiftIterators.v
+++ b/compiler/core/Utils/LiftIterators.v
@@ -15,8 +15,6 @@
 (** This module provides support for monadic iterators over bags. *)
 
 Require Import Arith.
-Require Import Min.
-Require Import Max.
 Require Import Lia.
 Require Import Permutation.
 Require Import Equivalence.
diff --git a/compiler/core/Utils/Monoid.v b/compiler/core/Utils/Monoid.v
index b3fa83c4..ca770152 100644
--- a/compiler/core/Utils/Monoid.v
+++ b/compiler/core/Utils/Monoid.v
@@ -56,8 +56,8 @@ Section Monoid.
         merge: A -> A -> A;
         z: A;
 
-        merge_morphism :> Proper (eqA ==> eqA ==> eqA) merge ;
-        z_morphism :> Proper (eqA) z;
+        #[global] merge_morphism :: Proper (eqA ==> eqA ==> eqA) merge ;
+        #[global] z_morphism :: Proper (eqA) z;
 
         merge_assoc: associative eqA merge;
         z_left: zero_left eqA merge z;
diff --git a/compiler/core/Utils/StringAdd.v b/compiler/core/Utils/StringAdd.v
index 9703f505..4e0e2e9b 100644
--- a/compiler/core/Utils/StringAdd.v
+++ b/compiler/core/Utils/StringAdd.v
@@ -20,7 +20,6 @@ Require Import Ascii.
 Require Import List.
 Require Import String.
 Require Import Arith.
-Require Import Min.
 Require Import Equivalence.
 Require Import EquivDec.
 Require Import Compare_dec.
@@ -381,8 +380,6 @@ Section Prefix.
     induction l; simpl; destruct s; destruct n; simpl; trivial.
     - f_equal.
       apply IHl.
-    - rewrite IHl; simpl; trivial.
-    - rewrite IHl. simpl; trivial.
   Qed.
 
   Lemma substring_all l :
@@ -401,8 +398,6 @@ Section Prefix.
       f_equal.
       f_equal.
       lia.
-    - rewrite IHl.
-      match_case.
   Qed.
 
   Lemma substring_le_prefix s n m l :
diff --git a/compiler/core/cNNRC/Lang/cNNRCVars.v b/compiler/core/cNNRC/Lang/cNNRCVars.v
index 96cb5a83..a75d0669 100644
--- a/compiler/core/cNNRC/Lang/cNNRCVars.v
+++ b/compiler/core/cNNRC/Lang/cNNRCVars.v
@@ -15,7 +15,6 @@
 Require Import String.
 Require Import List.
 Require Import Arith.
-Require Import Max.
 Require Import Peano_dec.
 Require Import EquivDec.
 Require Import Decidable.
diff --git a/compiler/core/cNRAEnv/Context/cNRAEnvContext.v b/compiler/core/cNRAEnv/Context/cNRAEnvContext.v
index 91d9774e..cff18a60 100644
--- a/compiler/core/cNRAEnv/Context/cNRAEnvContext.v
+++ b/compiler/core/cNRAEnv/Context/cNRAEnvContext.v
@@ -24,7 +24,6 @@ Require Import Program.
 Require Import List.
 Require Import Permutation.
 Require Import String.
-Require Import NPeano.
 Require Import Arith.
 Require Import Bool.
 Require Import Utils.
diff --git a/compiler/core/cNRAEnv/Context/cNRAEnvContextLift.v b/compiler/core/cNRAEnv/Context/cNRAEnvContextLift.v
index d62df334..729b06e2 100644
--- a/compiler/core/cNRAEnv/Context/cNRAEnvContextLift.v
+++ b/compiler/core/cNRAEnv/Context/cNRAEnvContextLift.v
@@ -23,7 +23,6 @@ Require Import EquivDec.
 Require Import Program.
 Require Import List.
 Require Import String.
-Require Import NPeano.
 Require Import Arith.
 Require Import Utils.
 Require Import DataRuntime.
diff --git a/compiler/core/cNRAEnv/Lang/cNRAEnv.v b/compiler/core/cNRAEnv/Lang/cNRAEnv.v
index da8aab12..c5cc2935 100644
--- a/compiler/core/cNRAEnv/Lang/cNRAEnv.v
+++ b/compiler/core/cNRAEnv/Lang/cNRAEnv.v
@@ -1210,7 +1210,7 @@ Section cNRAEnv.
       rewrite app_nil_r in *.
       rewrite lift_map_remove1 in *; simpl.
       rewrite (IHae1 env a); unfold nra_context_data; simpl.
-      generalize (h ⊢ (nra_of_nraenv_core ae1) @ₐ(drec
+      generalize (h ⊢ (nra_of_nraenv_core ae1) @ₐ (drec
             (("PBIND", env)
              :: ("PDATA", a) :: nil)) ⊣ constant_env); intros.
       destruct o; try reflexivity; simpl.
@@ -1319,7 +1319,7 @@ Section cNRAEnv.
       rewrite app_nil_r in *.
       rewrite lift_map_remove1 in *; simpl.
       rewrite (IHae1 env a); unfold nra_context_data; simpl.
-      generalize (h ⊢ (nra_of_nraenv_core ae1) @ₐ(drec
+      generalize (h ⊢ (nra_of_nraenv_core ae1) @ₐ (drec
             (("PBIND", env)
              :: ("PDATA", a) :: nil)) ⊣ constant_env); intros.
       destruct o; try reflexivity; simpl.
@@ -1378,7 +1378,7 @@ Section cNRAEnv.
     - Case "cNRAEnvDefault"%string.
       rewrite IHae1; rewrite IHae2; reflexivity.
     - Case "cNRAEnvEither"%string.
-      destruct x; simpl; trivial; [rewrite IHae1|rewrite IHae2]; reflexivity.
+      destruct x; simpl; trivial; solve [rewrite IHae1; reflexivity | rewrite IHae2; reflexivity].
     - Case "cNRAEnvEitherConcat"%string.
       rewrite IHae2. generalize ((h ⊢ (nra_of_nraenv_core ae2) @ₐ (nra_context_data env x) ⊣ constant_env)); intros.
       destruct o; try reflexivity; simpl;
diff --git a/compiler/core/cNRAEnv/Lang/cNRAEnvIgnore.v b/compiler/core/cNRAEnv/Lang/cNRAEnvIgnore.v
index c551f39d..4af34cd1 100644
--- a/compiler/core/cNRAEnv/Lang/cNRAEnvIgnore.v
+++ b/compiler/core/cNRAEnv/Lang/cNRAEnvIgnore.v
@@ -385,7 +385,7 @@ Section cNRAEnvIgnore.
       revert IHl.
       generalize (lift_filter
                     (fun x' : data =>
-                       match h ⊢ₑ e1 @ₑx' ⊣ c;env1 with
+                       match h ⊢ₑ e1 @ₑ x' ⊣ c;env1 with
                        | Some (dbool b) => Some b
                        | _ => None
                        end) l
diff --git a/compiler/core/cNRAEnv/Typing/TcNRAEnv.v b/compiler/core/cNRAEnv/Typing/TcNRAEnv.v
index c41edbac..bb5b1b56 100644
--- a/compiler/core/cNRAEnv/Typing/TcNRAEnv.v
+++ b/compiler/core/cNRAEnv/Typing/TcNRAEnv.v
@@ -367,7 +367,7 @@ Section TcNRAEnv.
                 data_type d (Rec Closed τ₁ pf1) ->
                 exists x : data,
                    brand_relation_brands ⊢ₑ op1 @ₑ y ⊣ c;env = Some x /\ data_type x (Coll (Rec Closed τ₂ pf2))) ->
-    exists x : list data, (omap_product (fun z =>  brand_relation_brands ⊢ₑ op1@ₑ y ⊣ c;env) dl = Some x) /\ data_type (dcoll x) (Coll (Rec Closed τ₃ pf3)).
+    exists x : list data, (omap_product (fun z =>  brand_relation_brands ⊢ₑ op1 @ₑ y ⊣ c;env) dl = Some x) /\ data_type (dcoll x) (Coll (Rec Closed τ₃ pf3)).
   Proof.
     intros dt_c Henv.
     intros; rewrite Forall_forall in *.
